sled-agent/src/vmm_reservoir.rs

```diff
@@ -0,0 +1,255 @@
+                }
+            }
+        }
+    }
+
+    /// Sets the VMM reservoir to the requested percentage of usable physical
+    /// RAM or to a size in MiB. Either mode will round down to the nearest
+    /// aligned size required by the control plane.
+    pub fn set_reservoir_size(
+        &self,
+        hardware: &sled_hardware::HardwareManager,
+        mode: ReservoirMode,
+    ) -> Result<(), Error> {
+        let hardware_physical_ram_bytes = hardware.usable_physical_ram_bytes();
+        let req_bytes = match mode {
+            ReservoirMode::None => return Ok(()),
+            ReservoirMode::Size(mb) => {
+                let bytes = ByteCount::from_mebibytes_u32(mb).to_bytes();
+                if bytes > hardware_physical_ram_bytes {
+                    return Err(Error::ReservoirConfig(format!(
+                        "cannot specify a reservoir of {bytes} bytes when \
+                        physical memory is {hardware_physical_ram_bytes} bytes",
+                    )));
+                }
+                bytes
+            }
+            ReservoirMode::Percentage(percent) => {
+                if !matches!(percent, 1..=99) {
+                    return Err(Error::ReservoirConfig(format!(
+                        "VMM reservoir percentage of {} must be between 0 and \
+                        100",
+                        percent
+                    )));
+                };
+                (hardware_physical_ram_bytes as f64 * (percent as f64 / 100.0))
+                    .floor() as u64
+            }
+        };
+
+        let req_bytes_aligned = vmm_reservoir::align_reservoir_size(req_bytes);
+
+        if req_bytes_aligned == 0 {
+            warn!(
+                self.log,
+                "Requested reservoir size of {} bytes < minimum aligned size \
+                of {} bytes",
+                req_bytes,
+                vmm_reservoir::RESERVOIR_SZ_ALIGN
+            );
+            return Ok(());
+        }
+
+        // The max ByteCount value is i64::MAX, which is ~8 million TiB.
+        // As this value is either a percentage of DRAM or a size in MiB
+        // represented as a u32, constructing this should always work.
+        let reservoir_size = ByteCount::try_from(req_bytes_aligned).unwrap();
+        if let ReservoirMode::Percentage(percent) = mode {
+            info!(
+                self.log,
+                "{}% of {} physical ram = {} bytes)",
+                percent,
+                hardware_physical_ram_bytes,
+                req_bytes,
+            );
+        }
+        info!(self.log, "Setting reservoir size to {reservoir_size} bytes");
+        vmm_reservoir::ReservoirControl::set(reservoir_size)?;
+
+        self.reservoir_size.store(reservoir_size.to_bytes(), Ordering::SeqCst);
+        info!(
+            self.log,
+            "Finished setting reservoir size to {reservoir_size} bytes"
+        );
+
+        Ok(())
+    }
+}
+
+impl VmmReservoirManager {
```

Comment:
nit: should these impl blocks be merged?

Reply:
Doh! Fixed in [bdea2f6](https://github.com/oxidecomputer/omicron/pull/5124/commits/bdea2f6c01707ad7b9323a9645f1e15e93e96104)